ehrapy/_utils_data.py

```diff
@@ -0,0 +1,53 @@
+from __future__ import annotations
+
+from typing import TYPE_CHECKING, Any
+
+import numpy as np
+
+if TYPE_CHECKING:
+    from anndata import AnnData
+
+
+def are_dataset_equal(arr1: np.ndarray, arr2: np.ndarray) -> np.bool_:
+    """Check if two arrays are equal member-wise.
+
+    Note: Two NaN are considered equal.
+
+    Args:
+        arr1: First array to compare
+        arr2: Second array to compare
+
+    Returns:
+        True if the two arrays are equal member-wise
+    """
+    return np.all(np.equal(arr1, arr2, dtype=object) | ((arr1 != arr1) & (arr2 != arr2)))
+
+
+def to_missing_data_matrix(data: np.ndarray) -> np.ndarray[Any, np.dtype[np.bool_]]:  # pragma: no cover
+    """Check if a AnnData value is missing.
+
+    Args:
+        data: The AnnData data to check
+
+    Returns:
+        An array of bool representing the missingness of the original data, with the same shape
+    """
+    return np.isin(data, [None, "", np.nan]) | (data != data)
+
+
+def to_dense_matrix(adata: AnnData, layer: str | None = None) -> np.ndarray:  # pragma: no cover
```

Comment:
The folks at scanpy have developed a very fast numba kernel for this which is much faster than scipy. See https://github.com/scverse/scanpy/blob/7131500627a3037cedd11d380ae772400c744f9b/src/scanpy/preprocessing/_utils.py#L165

I have already advocated for this to be public API in anndata instead so that all packages can make use of it. @ilan-gold @Intron7 .

Reply:
I don't think that we can rely on it being where it currently is. I'll lobby for it to be public and then we can start using it. For now, let's just use what we have.